sycl/include/CL/sycl/kernel_bundle.hpp

```diff
@@ -259,20 +266,39 @@ class kernel_bundle : public detail::kernel_bundle_plain {
             typename = detail::enable_if_t<_State == bundle_state::input>>
   void set_specialization_constant(
       typename std::remove_reference_t<decltype(SpecName)>::value_type Value) {
-    (void)Value;
-    throw sycl::runtime_error(
-        "kernel_bundle::set_specialization_constant is not implemented yet",
-        PI_INVALID_OPERATION);
+
+    using SCType =
+        typename std::remove_reference_t<decltype(SpecName)>::value_type;
+    static_assert(std::is_trivially_copyable_v<SCType>);
+    // TODO can this be simply default constructible
+    static_assert(std::is_trivially_default_constructible_v<SCType>);
+
+    const char *SpecSymName = detail::get_spec_constant_symbolic_ID<SpecName>();
+    set_specialization_constant_impl(SpecSymName, &Value, sizeof(SCType));
   }
 
-  /// The value of the specialization constant whose address is SpecName for
-  /// this kernel bundle.
+  /// \returns the value of the specialization constant whose address is
+  /// SpecName for this kernel bundle.
   template <auto &SpecName>
   typename std::remove_reference_t<decltype(SpecName)>::value_type
   get_specialization_constant() const {
-    throw sycl::runtime_error(
-        "kernel_bundle::get_specialization_constant is not implemented yet",
-        PI_INVALID_OPERATION);
+    using SCType =
+        typename std::remove_reference_t<decltype(SpecName)>::value_type;
+    static_assert(std::is_trivially_copyable_v<SCType>);
+    // TODO can this be simply default constructible
+    static_assert(std::is_trivially_default_constructible_v<SCType>);
+
+    if (!has_specialization_constant<SpecName>())
+      // TODO replace with SYCL 2020 exception
+      throw sycl::runtime_error("Unknown specialization constant",
+                                PI_INVALID_VALUE);
```

Comment:
It seems like spec does not really clarify, what happens if this spec constant does not present in kernel bundle. Tagging @AlexeySachkov.

Reply:
I'm not exactly sure, but from [this section](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_setting_and_getting_the_value_of_a_specialization_constant):

> This function may be called even if the specialization constant SpecName isn’t used by the kernel that is invoked by this handler’s command group. Doing so has no effect on the invoked kernel.

It seems like it should be ok to set a constant which doesn't exists. It also seems a bit error-prone, because users might wonder what went wrong, while they are just set wrong specialization constant and didn't get any diagnostic about that. Tagging @gmlueck here for more comments

Reply:
I should correct myself: I missed the fact that we are talking about `kernel_bundle`, then the correct spec section would be a [this one](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_specialization_constant_support), but it says similar thing:

> The new value applies to all device images in the bundle. It is allowed to set the value of a specialization constant even if no device image in the bundle uses it; doing so has no effect on the execution of kernels from that bundle.

Reply:
@AlexeySachkov let me clarify a bit. We're talking about `get_specialization_constant`. If I have a code like this:
```c++
void foo(sycl::kernel_bundle KB) {
  auto Val = KB.get_specialization_constant<SpecName>();
}
```
What should happen if this `KB` doesn't have `SpecName`? The spec reads the following:

> Returns: The value of the specialization constant whose address is SpecName for this kernel bundle. The value returned is as follows:
> - If the value of this specialization constant was previously set in this bundle, that value is returned. Otherwise,
> - If this bundle is the result of compiling, linking or joining another bundle and this specialization constant was set in that other bundle prior to compiling, linking or joining; then that value is returned. Otherwise,
> - The specialization constant’s default value is returned.

Does it mean that we simply return the default value?

Reply:
> Does it mean that we simply return the default value?

@alexbatashev, it seems so

Reply:
No, the spec does not allow a runtime exception to be thrown in this case.

The `kernel_bundel::set_specialization_constant()` and `kernel_bundle::get_specialization_constant()` work the same regardless of whether there is any reference to the particular specialization constant in the kernel bundle.  You can still set the value even if there is no reference in the kernel bundle, and you can still read the value even if there is no reference in the kernel bundle.

Reading the value of a spec constant that is not referenced returns whatever value was previously set in that bundle.  If no value was previously set, you get the default value.

I think this is not as error prone as you suggest.  The compiler will give an error if you try to get or set the value of a specialization constant that was never declared.  This is only a question of what happens if the declared spec constant isn't **used** in this particular kernel bundle.

I wanted this behavior because the application doesn't have full control over the set of spec constants that are used in a bundle.  Consider a case where the application asks for a bundle that contains kernel `Foo`.  The implementation is allowed to return a bundle that includes other kernels `Bar` and `FooBar` also.  In such a case, the bundle will include any spec constants used in `Bar` and `FooBar` also, even though they are not used in `Foo`.

Now consider a case where this same application calls `kernel_bundle::set_specialization_constant()` to set the value of some spec constant that is used only in `Bar`.  The implementation can't diagnose this as an error because `Bar` is contained by the bundle (even though the application didn't request a bundle with `Bar`).  Now, consider this same application running on some other implementation that happens to bundle kernels differently.  In this second implementation, the application gets a bundle without `Bar`.  I wanted to avoid a case where the application gets an exception on one implementation but not on another.